[torch.Package/TorchScript] TS into torch.Package test cases

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57888 [PyTorch Edge] bytecode version bump to v5 and enable share constant table
* #58182 [PyTorch Edge] share tensors in mobile with new api
* #57678 [torch.package/TorchScript] flag to gate allowance of TS serializaiton in torch.package
* **#54894 [torch.Package/TorchScript] TS into torch.Package test cases**
* #57573 [torch.package/TorchScript] logic to enable sharing of tensors on load
* #54893 [torch.Package/TorchScript] torch.Package python logic to save TorchScript
* #56299 [torch.Package/TorchScript] ScriptModuleSerializer add unified format
* #55958 [torch.Package/TorchScript] refactor ScriptModuleSerializer Exporter
* #54891 [torch.Package/TorchScript] TS serialization importer to handle unified format

Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

Commit:
[torch.Package/TorchScript] TS into torch.Package test cases

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

suo reviewed (changes requested):
flipping back to you, feel free to re-request when ready

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented on test/package/test_torchscript.py:
```diff
@@ -0,0 +1,418 @@
+    def test_save_ts_modules_shared_code(self):
+        #  Test to verify saving multiple modules with same top module
+        #  but different submodules works
+        class ModD(torch.nn.Module):
```
Should I duplicate all the tests to use the inline ones? Or should just one or two be enough? The last test in this file does load from package_a

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented on test/package/test_torchscript.py:
```diff
@@ -0,0 +1,576 @@
+        self.assertEqual(loaded_mod_list[0]("input"), scripted_mod_a("input"))
+        self.assertEqual(loaded_mod_list[1]("input"), scripted_mod_b("input"))
+
+    def test_save_shared_scriptmodules(self):
```
Here's the current behavior of shared ScriptModules

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented on test/package/test_torchscript.py:
```diff
@@ -0,0 +1,418 @@
+        loaded_mod_scripted = importer_1.load_pickle("res", "scripted_mod.pkl")
+
+        self.assertTrue(torch.allclose(loaded_mod_scripted(input), orig_mod(input)))
+
```
I added tests to confirm both of these things (test_save_shared_scriptmodules and test_save_shared_tensors)

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
Changes since last review:
- added tests to verify sharing of tensors and sharing of scriptmodules
- added tests which mix imported and inline scripted modules 
- cleaned up doc strings and naming

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

suo commented on test/package/test_torchscript.py:
```diff
@@ -0,0 +1,584 @@
+        importer = PackageImporter(buffer)
+        file_structure = importer.file_structure()
+        self.assertTrue(file_structure.has_file(".data/ts_code/0"))
+        self.assertFalse(file_structure.has_file(".data/ts_code/1"))
```
We want to check two things:
1. On save, we only serialize once
2. On load, we only deserialize once (e.g. we share the same ScriptModule object across all three loaded modules)

Right now we're checking 1), but let's check 2) as well. We can add an identity assertion: if we load_pickle the three mods, the `id` of their script_mod attribute should be the same, right? Or maybe the Python wrapper is different, but the underlying C impl should be the same.

Lilyjjo replied:
I think I'm confused about how pickled object inside of a package are supposed to relate to each other. Are the relationships between objects supposed to be upheld between `load_pickle`s? I can understand preserving the relationships between objects that are contained in a parent object that is loaded from a pickle, but not between `load_pickle`s. The current `PackageImporter.load_pickle` reads to me like it only preserves the relationship between storages within a `load_pickle`-- in other words each `load_pickle` will re-load the storages newly each time. We could implement a cache on the importer to allow storages to be loaded only once, and for script-modules, but idk if that matches the normal pickle behavior. If a user pickle_loads the same pickle twice, should they get the same object by python `id` twice? We don't do that currently

Lilyjjo replied:
The real thing I'm confused about is if `load_pickle` is supposed to return a new instance of the saved object each time, or if it's always supposed to return the same instance. Like if you call `load_pickle("res", "mod1.pkl")` twice, should both returned objects be the same instance? Right now they're not:
```
''' Mod definition:
class Mod(torch.nn.Module):
    def __init__(self, x, y):
        super().__init__()
        self._list = x
        self._tensor = y

    def forward(self, x):
        return 
'''

mod = Mod([], torch.rand(1,2,3))

buffer = BytesIO()
with PackageExporter(buffer, verbose=False) as e:
    e.extern(include=["torch", "sys", "io"])
    e.save_pickle("model", "model.pkl", mod)

buffer.seek(0)
importer = PackageImporter(buffer)

loaded_mod1 = importer.load_pickle("model", "model.pkl")
loaded_mod1_repeat = importer.load_pickle("model", "model.pkl") 

assert id(loaded_mod1) != id(loaded_mod1_repeat)
assert id(loaded_mod1._tensor) != id(loaded_mod1_repeat._tensor)

assert torch.allclose(loaded_mod1._tensor, loaded_mod1_repeat._tensor) == 1
loaded_mod1._tensor = loaded_mod1._tensor * 2
assert torch.allclose(loaded_mod1._tensor, loaded_mod1_repeat._tensor) == 0
```

suo replied:
Ah sorry I think my suggestion was wrong. You're right: each call to `load_pickle` should produce a new copy of everything.

I was thinking of like a diamond shape:
```
        parent
       /  |  \
      A   B   C
      \   |   /
       child
```

if you do `load_pickle(parent)`, it should be the case that `a.A.child is a.B.child is a.C.child`, where child can either be a tensor or a scriptmodule.

Lilyjjo replied:
Awesome. The code now preserves this relationship on load for both ScriptModules and storages. The tests to show them have '_load' in the name of the test

suo commented on test/package/test_torchscript.py:
```diff
@@ -0,0 +1,584 @@
+        importer = PackageImporter(buffer)
+        loaded_mod_1 = importer.load_pickle("res", "mod1.pkl")
+
+        # assert that there is only one storage stored in package
```
same feedback as above. We want to ensure that we only loaded the relevant storages once.

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Lilyjjo commented on test/package/test_torchscript.py:
```diff
@@ -0,0 +1,626 @@
+        IS_FBCODE or IS_SANDCASTLE,
+        "Tests that use temporary files are disabled in fbcode",
+    )
+    def test_save_scriptmodule_file(self):
```
This test is for testing to make sure TS works in a file instead of in just a buffer. The test below it needs a file also to create the directory out of (the rest of the tests all use buffers)

suo replied:
ah I see, cool

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832544).

Commit:
Update on "[torch.Package/TorchScript] TS into torch.Package test cases"


Test cases to test torch.Package's handling of TorchScript objects.

TODO: test mapping storages to different device

Differential Revision: [D27832544](https://our.internmc.facebook.com/intern/diff/D27832544)

[ghstack-poisoned]
